### PR TITLE
Full support for custom service cluster ip range

### LINKED ISSUE #1536: Dashboard does not launch with different service cluster IP range
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Minikube version** (use `minikube version`): v0.19.0

**Environment**:
- **OS** (e.g. from /etc/os-release): Ubuntu 17.04
- **VM Driver** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep DriverName`): VirtualBox 
- **ISO version** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep -i ISO` or `minikube ssh cat /etc/VERSION`): v0.18.0
- **Install tools**: None
- **Others**: None

**What happened**: I try to change the service cluster IP range (which works) using the `--extra-config` command line option suggested [here](https://github.com/kubernetes/minikube/issues/1323#issuecomment-293606503). However, with a clean setup, it does not launch the dashboard:

```
$ minikube dashboard
Waiting, endpoint for service is not ready yet...
...
...
Waiting, endpoint for service is not ready yet...
Could not find finalized endpoint being pointed to by kubernetes-dashboard: Temporary Error: No endpoints for service are ready yet
Temporary Error: No endpoints for service are ready yet
...
...
Temporary Error: No endpoints for service are ready yet
Error validating service: Error getting service kubernetes-dashboard: Get https://192.168.99.100:8443/api/v1/namespaces/kube-system/services/kubernetes-dashboard: dial tcp 192.168.99.100:8443: getsockopt: connection refused
```

Also, running `kubectl proxy` and navigating to `http://localhost:8001/ui` shows the following:

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "no endpoints available for service \"kubernetes-dashboard\"",
  "reason": "ServiceUnavailable",
  "code": 503
}
```

**What you expected to happen**:

The dashboard should open.

**How to reproduce it** (as minimally and precisely as possible):

`minikube stop && minikube delete && minikube start --extra-config=apiserver.ServiceClusterIPRange=172.66.0.0/24`

When ready, run `minikube dashboard`.

**Anything else do we need to know**:

It does work when I create a clean cluster using only `minikube start`.

### LINKED ISSUE #1747: Critical errors when apiserver.ServiceClusterIPRange has been modified
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG

**Minikube version** (use `minikube version`): v0.20.0

**Environment**:
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 8 (jessie)
- **VM Driver** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep DriverName`): "virtualbox"
- **ISO version** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep -i ISO` or `minikube ssh cat /etc/VERSION`): v0.20.0
- **Install tools**:
- **Others**: Docker version 17.06.0-ce, build 02c1d87

**What happened**:
```
$ minikube start --extra-config=apiserver.ServiceClusterIPRange=10.10.240.0/24
$
$ kubectl get svc
NAME         CLUSTER-IP    EXTERNAL-IP   PORT(S)   AGE
kubernetes   10.10.240.1   <none>        443/TCP   3m
$
$ kubectl cluster-info dump   ## Extract
The Service "kube-dns" is invalid: spec.clusterIP: Invalid value: "10.0.0.10": provided IP is not in the valid range
E0725 18:06:59.172788       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Service: Get https://10.10.240.1:443/api/v1/services?resourceVersion=0: x509: certificate is valid for 192.168.99.100, 10.0.0.1, not 10.10.240.1
E0725 18:06:59.173134       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: Get https://10.10.240.1:443/api/v1/endpoints?resourceVersion=0: x509: certificate is valid for 192.168.99.100, 10.0.0.1, not 10.10.240.1
```

- kube-dns service cannot start.
- The API certificates are invalid for current IP address (10.10.240.1). 

**What you expected to happen**:
Minikube starts successfully when apiserver.ServiceClusterIPRange option has been modified.

**How to reproduce it** (as minimally and precisely as possible):
```
$ minikube start --extra-config=apiserver.ServiceClusterIPRange=10.10.240.0/24
```

**Anything else do we need to know**:
No

### PR DESCRIPTION
**Pull request by rodcloutier**
This PR intends to add full support for custom service cluster ip range.

Fixes #1536 
Fixes #1747 

Feel free to add some more requirements for testing.

### CODE REVIEW

**Comment by dlorenc** [comment]
@minikube-bot ok to test

File: pkg/minikube/cluster/cluster.go [hunk 1]
```diff
@@ -173,7 +172,7 @@ func GetHostDriverIP(api libmachine.API) (net.IP, error) {
 func engineOptions(config MachineConfig) *engine.Options {
 	o := engine.Options{
 		Env:              config.DockerEnv,
-		InsecureRegistry: append([]string{pkgutil.DefaultServiceCIDR}, config.InsecureRegistry...),
+		InsecureRegistry: append([]string{config.ServiceCIDR}, config.InsecureRegistry...),
```
Comment (bacongobbler) [inline pkg/minikube/cluster/cluster.go]: :+1: glad this was picked up in the PR :)

**Comment by jar349** [comment]
/remove-lifecycle stale

I don't believe that this PR should close due to inactivity when it offers a point of configurability that would - at least to me - be useful.  My use case is this: https://stevesloka.com/2017/05/19/access-minikube-services-from-host/

**Comment by jar349** [comment]
this one even includes a patch!  it ain't rotten.

**Comment by jar349** [comment]
@rodcloutier I wonder if it would be possible for you to freshen this up to make it super-simple for admins to approve?

**Comment by dsiebel** [comment]
I just built the minikube.iso image from this branch hoping it could potentially fix my issues with colliding network ranges (company networks spanning 172.16/12 and 10.0/8) but sadly it didn't. The issue was with routing service IPs (10.96/12) within the cluster, which always went through my host and was terminated by our company firewall.

I tried starting minikube with the following settings but it just stalls and aborts after 10 minutes or so:
```
minikube start --logtostderr --v=3 \
  --cpus 4 \
  --memory 4096 \
  --keep-context \
  --docker-opt="max-concurrent-downloads=8" \
  --docker-opt="bip=192.168.5.1/24" \
  --extra-config="apiserver.Authorization.Mode=RBAC" \
  --extra-config="apiserver.ServiceClusterIPRange=192.168.48.0/20" \
  --extra-config="kubelet.PodCIDR=192.168.32.0/20" \
  --extra-config="controller-manager.ClusterCIDR=192.168.32.0/20"
```
I also tried removing single CIDR parameters but the result didn't change.

In summary: not sure whether this PR actually fixes the issue but it would be really great if it did 😃

**Comment by tstromberg** [comment]
Do you mind addressing the merge conflicts? I'd like to see about merging this if it's still viable. My apologies for the unnecessarily long delay in reviewing this PR.

**Comment by tstromberg** [comment]
Any hope for resurrecting this PR from the merge conflicts? I feel like it's doing the right thing and would love to see it in minikube. Thanks for all the hard work you've put into it so far.

**Comment by rodcloutier** [comment]
I will see what I can do.

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 2]
```diff
@@ -294,15 +294,22 @@ func NewKubeletConfig(k8s config.KubernetesConfig, r cruntime.Manager) (string,
 		return "", errors.Wrap(err, "parses feature gate config for kubelet")
 	}
 
+	clusterDNS, err := util.GetDNSIP(k8s.ServiceCIDR)
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: Can you mention why the DNS changes are necessary in the PR comment?

File: pkg/minikube/config/types.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ type MachineConfig struct {
 	InsecureRegistry    []string
 	RegistryMirror      []string
 	HostOnlyCIDR        string // Only used by the virtualbox driver
+	ServiceCIDR         string // Use for insecure registry
```
Comment (tstromberg) [inline pkg/minikube/config/types.go]: I don't understand why the ServiceCIDR is only to be used with the insecure registry. Can you clarify this comment?

**Comment by tstromberg** [comment]
Thank you for the contribution. Due to recent inactivity and failing tests, I am marking this PR as closed until the issues can be resolved.

I really would like to see this PR come into being, so please re-open once it seems to be working. Before doing so, please run the integration tests locally: https://github.com/kubernetes/minikube/blob/master/docs/contributors/build_guide.md#integration-tests

 Thank you again!
